truffle/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/InstrumentationHandler.java

```diff
@@ -0,0 +1,716 @@
+        ProbeNode probeNode = parent.getProbeNode();
+        if (TRACE) {
+            SourceSection section = probeNode.getContext().getInstrumentedSourceSection();
+            trace("Invalidate wrapper for %s, section %s tags %s%n", node, section, Arrays.toString(section.getTags()));
+        }
+        if (probeNode != null) {
+            probeNode.invalidate();
+        }
+    }
+
+    static Instrumentable getInstrumentable(Node node) {
+        Instrumentable instrumentable = node.getClass().getAnnotation(Instrumentable.class);
+        if (instrumentable != null && !(node instanceof WrapperNode)) {
+            return instrumentable;
+        }
+        return null;
+    }
+
+    private abstract class AbstractNodeVisitor implements NodeVisitor {
+
+        abstract boolean shouldVisit(RootNode root);
+
+    }
+
+    private abstract class AbstractBindingVisitor extends AbstractNodeVisitor {
+
+        protected final EventBinding<?> binding;
+
+        AbstractBindingVisitor(EventBinding<?> binding) {
+            this.binding = binding;
+        }
+
+        @Override
+        boolean shouldVisit(RootNode root) {
+            return isInstrumentedRoot(root, binding, root.getSourceSection());
+        }
+
+        public final boolean visit(Node node) {
+            SourceSection sourceSection = node.getSourceSection();
+            if (sourceSection != null) {
+                if (isInstrumentedLeaf(node, binding, sourceSection)) {
+                    if (TRACE) {
+                        trace("Filter hit section:%s tags:%s%n", sourceSection, Arrays.toString(sourceSection.getTags()));
+                    }
+                    visitInstrumented(node, sourceSection);
+                }
+            }
+            return true;
+        }
+
+        protected abstract void visitInstrumented(Node node, SourceSection section);
+
+    }
+
+    private abstract class AbstractBindingsVisitor extends AbstractNodeVisitor {
+
+        private final List<EventBinding<?>> bindings;
+
+        AbstractBindingsVisitor(List<EventBinding<?>> bindings) {
+            this.bindings = bindings;
+        }
+
+        @Override
+        boolean shouldVisit(RootNode root) {
+            SourceSection sourceSection = root.getSourceSection();
+            for (int i = 0; i < bindings.size(); i++) {
+                EventBinding<?> binding = bindings.get(i);
+                if (isInstrumentedRoot(root, binding, sourceSection)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        public final boolean visit(Node node) {
+            SourceSection sourceSection = node.getSourceSection();
+            if (sourceSection != null) {
+                List<EventBinding<?>> b = bindings;
+                for (int i = 0; i < b.size(); i++) {
+                    EventBinding<?> binding = b.get(i);
+                    if (isInstrumented(node, binding, sourceSection)) {
```

Comment:
Ok, so I was hitting this issue that `RootNode` objects get instrumented, even so they are not carrying the @Instrumentable annotation. Part of this problem seems to be this `isInstrumented()` check. 

[Also, I am not sure the name is really what is intended. This sounds like as if the node has been instrumented in the past/is already instrumented, which is not the way the visitor is used.]

Hacking in an `&& !(node instanceof RootNode)` works for the moment, but does not look like the right thing to do, especially since I am not sure in which context this visitor is supposed to be used.
Depending on what the semantics of this `isInstrumented` method are supposed to be, it might need to check whether a node has an annotation at the moment, it only really checks the source section. It also does a `isInstrumentable(node)` check, but for instance `InstrumentationInstrumenter` returns just `true`.

I didn't really manage to adapt a test for this purpose either. The existing ones in `InstrumentationTest` don't seem easily coerced into the issue, because of all kind of other complex interplay of conditions. (I tried to mark the DEFINE node also as ROOT, and EXPRESSION, because there is filtering going on, and using this source section in the `InstrumentationTestRootNode`, but without getting the same exception.